Implemented TwitterSourceFilter and adapted TwitterSource

HilmiYildirim opened the pull request:
[FLINK-2048] I implemented the class TwitterFilterSource which inherits from TwitterSource. It enables to filter twitter streams, for example, by terms, languages and locations. Furthermore, you can add "followings" to the stream by adding user ids to the config. Then, the stream includes the tweets of these persons. Furthermore, it is possible to define query and post parameters.

I had to make some changes in the TwitterSource file. I had to change some variables and methods from private to protected. Furthermore, now TwitterSource inherits from RichSourceFunction instead of RichParallelSourceFunction. RichParallelSourceFunction does not make sense. With a twitter handle it is only possible to open 2 twitter streams in parallel. But if TwitterSource inherits from RichSourceFunction then only one stream will be opened and everything works fine.

rmetzger commented:
Thank you for working on the issue. I'll soon review your changes in detail.
From what I saw by scrolling over it, it looks very good (with log messages and documentation).
One thing that caught my attention was that the included test is not really a unit or integration test. Its more like an example.
Do you think there is a way to turn this into a real JUnit test?

HilmiYildirim commented:
It is possible to implement an integration test. For example, I can 
track a term and then I has to check if all tweets contain the term. If 
you want I can implement such tests.

Best Regards,
Hilmi

Am 19.05.2015 um 15:39 schrieb Robert Metzger:
>
> Thank you for working on the issue. I'll soon review your changes in 
> detail.
> From what I saw by scrolling over it, it looks very good (with log 
> messages and documentation).
> One thing that caught my attention was that the included test is not 
> really a unit or integration test. Its more like an example.
> Do you think there is a way to turn this into a real JUnit test?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/apache/flink/pull/695#issuecomment-103498471>.
>

-- 
--
Hilmi Yildirim
Software Developer R&D

T: +49 30 24627-281
Hilmi.Yildirim@neofonie.de

http://www.neofonie.de

Besuchen Sie den Neo Tech Blog für Anwender:
http://blog.neofonie.de/

Folgen Sie uns:
https://plus.google.com/+neofonie
http://www.linkedin.com/company/neofonie-gmbh
https://www.xing.com/companies/neofoniegmbh

Neofonie GmbH | Robert-Koch-Platz 4 | 10115 Berlin
Handelsregister Berlin-Charlottenburg: HRB 67460
Geschäftsführung: Thomas Kitlitschko

rmetzger commented:
That would be really great, thank you!

hsaputra commented on flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/twitter/TwitterFilterSource.java:
```diff
@@ -0,0 +1,284 @@
+ * An extension of {@link TwitterSource} by filter parameters. This extension
+ * enables to filter the twitter stream by user defined parameters.
+ * 
+ * @author Hilmi Yildirim
```
Could you remove @author javadoc annotation?

HilmiYildirim replied:
Yes I can remove the annotation.

There is a problem with the junit tests. For them we need a twitter account
and the access keys for the stream connection have to be in the resource
folder. This means that other people can use the access keys which can
influence the tests. I think there is no secure way to implement junit
tests.

hsaputra commented on flink-staging/flink-streaming/flink-streaming-connectors/src/test/java/org/apache/flink/streaming/connectors/twitter/TwitterFilterSourceTest.java:
```diff
@@ -0,0 +1,56 @@
+
+	private static final String PATH_TO_AUTH_FILE = "/twitter.properties";
+	
+	public static void main(String[] args) {
```
We are using JUnit for our tests so it would be great to convert this to JUnit style of test class.

HilmiYildirim replied:
Unfortunately, it is not possible it implement this as Junit Tests. The stream connection needs access keys of a twitter account and the maximum number of parallel streams per twitter account is restricted to 2. If more than 2 people are executing the tests at the same time then the tests fail.

HilmiYildirim replied:
But it would be possible to implement this as Junit tests and the tests will be skipped by default. Someone who wants to execute the tests has to use his own access keys.

rmetzger replied:
I see. The source is expecting a running twitter service.
Our tests should be self-contained, without relying on external services.
For the Kafka connector, we are starting an internal testing kafka server in the same JVM and test Flink-kafka against that one.

Do you think its doable to implement a little server which is behaving like twitters APIs ?
That would be the very best way to test this reliably.
Maybe it would be sufficient to make `initializeClient()` protected, create a `TestingTwitterSource` which is overriding `initializeClient()` and creating a testing implementation of `com.twitter.hbc.core.Client` ?

But I also understand that its probably more work to implement that test compared to your contribution.
Since we didn't have tests before, I would also agree to merge your changes now and file a issue for implementing the test.

HilmiYildirim replied:
I think there is a problem of overwriting com.twitter.hbc.core.Client. The stream filtering process is done by the twitter server. The filter parameters (terms, languages, ...) are sent to the twitter server to establish the twitter stream. Then, the twitter server only sends tweets to the client which fit the filtering criteria. Therefore, if I overwrite com.twitter.hbc.core.Client then I have to implement the whole filtering mechanism which leads to, that the Junit tests test the "test implementation" instead of the "real implementation".

rmetzger replied:
I agree, lets not add a test for now.

hsaputra replied:
Sounds good to me

HilmiYildirim commented:
Unfortunately, it is not possible ti implement this as Junit Tests. The stream connection needs access keys of a twitter account and the maximum number of parallel streams per twitter account is restricted to 2. If more then 2 people are executing the test at the same time then the tests are not working.

rmetzger commented:
+1 to merge like this.

hsaputra commented on flink-staging/flink-streaming/flink-streaming-connectors/src/test/java/org/apache/flink/streaming/connectors/twitter/TwitterFilterSourceTest.java:
```diff
@@ -0,0 +1,56 @@
+import org.apache.flink.streaming.connectors.json.JSONParseFlatMap;
+import org.apache.flink.util.Collector;
+
+public class TwitterFilterSourceTest {
```
Could you add JavaDoc comment on what is needed to run this test and kind of small description why it is not part of JUnit test for now.

hsaputra commented:
Add one more small request to add comments as JavaDoc for the test then it should ready to go.
Thanks again for the contribution.

HilmiYildirim commented:
I renamed the TwitterFilterSourceTest to TwitterFilterSourceExample because it is more an example instead of a test. Further, I added JavaDoc comments and I also added a twitter.properties dummy to the resource folder that you only has to fill the twitter.properties to execute the example.

HilmiYildirim commented:
TwitterSource does not work after the last update. Now, it sends null messages. I try to fix it.

HilmiYildirim commented:
The problem was that the TwitterSource.next method was not implemented properly. There were also dead code. I fixed it and now it is working.

rmetzger commented:
It looks like many unrelated commits are part of the pull request. I suspect something went wrong while merging/rebasing.
Can you take a look?

HilmiYildirim commented:
The problem is as follows. While I was working on the commit, the interface SourceFunction has changed and was merged with the master. Now, it has the methods reachedEnd and next. Cause TwitterSource inherits from SourceFunction, the nextReachedEnd and next methods were implemented in the class. The problem os that the implementation has bugs and the old methods are not deleted which leads to that the class also has dead code. Therefore, the TwitterFilterSource class was not working and, therefore, TwitterFilterSource was also not working. Therefore, I have fixed the class TwitterSource and removed the old code. Now, it is working properly.

StephanEwen commented:
I tried to merge this, but it is quite unmergeable.

Before we can add this, we need you to rebase this pull request and squash all commits into one commit.

HilmiYildirim commented:
Done. I hope every thing works fine.

aljoscha commented:
Is this good to merge now? I'm rewriting all the sources (again) and this should probably go in.

HilmiYildirim commented:
I think it is good.

aljoscha commented:
Thanks for your work. :smile:
